src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerActor.scala

```diff
@@ -266,13 +268,30 @@ private[impl] class GroupManagerActor(
 
     val dynamicApps: Set[AppDefinition] =
       to.transitiveApps.map {
-        // assign values for service ports that the user has left "blank" (set to zero)
-        case app: AppDefinition if app.hasDynamicServicePorts => assignPorts(app)
-        case app: AppDefinition =>
-          // Always set the ports to service ports, even if we do not have dynamic ports in our port mappings
+        case app: AppDefinition if app.servicePorts.nonEmpty =>
+          synchronized {
+            app.servicePorts.withFilter(_ != AppDefinition.RandomPortValue).foreach { port => taken += port }
+          }
+          assignPorts(app)
+        case app: AppDefinition if app.portDefinitions.exists(_.port == AppDefinition.RandomPortValue) =>
```

Comment:
How so? These are "portDefintion" style ports.

The way servicePorts are implemented today and I _still_ think its completely broken (e.g. 1.3) - is that they are _globally unique_ not "servicePort: A service port is a port used to describe the port that a service should made available at. Marathon does not bind to the service ports specified but will ensure that you cannot have multiple applications that use the same service port running on the same host. Service ports are typically only used by external applications (e.g. HAProxy) to make the application available at the specified port. See Service Discovery & Load Balancing for more information."

This is the major regression, so maybe this entire suite of logic should be killed off and the old behavior should be brought back? (e.g. no servicePort uniqueness)